src/util/numbering.h

```diff
@@ -18,53 +18,42 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <util/invariant.h>
 #include <util/optional.h>
 
-template <typename T>
-// NOLINTNEXTLINE(readability/identifiers)
-class numbering final
+/// \tparam Key a value type
+/// \tparam Map a metafunction which is invoked with two parameters to
+/// return an appropriate mapping class
+template <class Key, class Map>
+class template_numberingt final
```

Comment:
Why not make this a template taking a single parameter, which would be the fully qualified map type? You could then use `T::key_type` to pick out the key type rather than explicitly passing it in.

But maybe I'm wrong and there's a reason this cannot be done.

Reply:
You end up duplicating info, or with possible type mismatches. We want the number_type to be specified in the class (here as a size_t). Making the user pass in a map from `key` to `size_t` implies that the user has control over the mapped-to type, even though they really don't.

Reply:
Good point, but how about addressing this via `static_assert` and documentation? I do agree that it doesn't seem perfect either, but at least it seems much simpler to me than the current proposal.

Reply:
In general it's better to make the requirements of the code explicit in the code itself, rather than to rely on documentation (which tends to get out-of-sync with code over time). Allowing the user to specify a numeric type is either redundant, if the type is correct, or leads to a surprising error when the types are mismatched. Therefore I believe the current solution is preferable.

Reply:
Basically, simple with incorrect/misleading semantics is worse than complicated but with the correct semantics.

Reply:
I might add that the type is unnecessarily fixed. Any type that can represent the number of objects that the user wants numbered is actually good enough. I do agree that correct semantics are important, but as long as we can enforce them via a `static_assert` that's perfectly sufficient.

Reply:
OK, if we're saying that the numbering type can change too, then this approach seems reasonable.

Reply:
Done